wine 3.6 (devel)

LinusU opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This was briefly discussed in #23614, but I think it's appropriate to open again.

Of course, bumping the build time from 1h to 2h is unfortunate, but since the now stable version (3.0.x) will only be updated every 3 months, and the devel version every two weeks, I don't think it's too bad.

Benefits:
- [their FAQ recommends trying the development releases](https://wiki.winehq.org/FAQ#Which_version_of_Wine_should_I_use.3F)
- 3.0 is already missing some great features, such as the Direct3D multi-threaded command stream
- many bugs get fixed continually but takes up to 3 months to reach stable, if they do at all

fxcoudert commented:
poke @ilovezfs

ilovezfs commented:
I'm willing to add this if the patch is removed. We cannot keep patching wine forever.

LinusU commented:
> I'm willing to add this if the patch is removed.

I would be very sad to see the patch go, without it, a ton of games becomes unplayable on macOS.

It's my understanding, from being a part of that thread for a long time, that the wine developers think that this should be fixed in Apples OpenGL driver. I personally don't think that that's going to happen at all.

> We cannot keep patching wine forever.

I understand that it's annoying to keep patches, and the wine formula is already super long compared to other formulas. But I think that what makes this a bit different is that upstream doesn't seem that concerned with macOS support, and Homebrew is a package manager for macOS.

> I'm afraid I'll have to disappoint, it's really up to Apple to fix this. There may be architectural changes that may mitigate the issue in the future, but I can't promise that.
> 
> The thing is that the Apple OpenGL implementation has many more bugs and quirks than just this one, and while the individual workarounds for those may seem innocent enough on their own, those quickly add up to something unmaintainable. The other consideration is that Wine is primarily a Free Software project, and has comparatively few users and developers on MacOS. That means that even if we did include those kinds of workarounds, it would be relatively easy to inadvertently break them again.

From this ☝️ it seems to me that they simply do not care that much about macOS support...

If there is anything that would help with maintaining the patch, I would be happy to do so. I've rebased it onto newer versions before and would be happy to do so again. Maybe we could add a comment to ping me here on Github if the patch fails with a newer version, that way I could provide a new patch as soon as possible.

This patch is making me and at least a few people I know play some great classics (Age of Empires & RollerCoaster Tycoon anyone? 🙌) on the macOS, I would be sad to see it go...

ilovezfs commented:
@LinusU it may require forking the wine project into a Mac specific version.

fxcoudert commented:
@adunning The website clearly calls them “stable” and “development”, so no

albinekb commented:
I would like to see this land. It would be awesome to be able to play classics on the Mac. As far as I can see the "patch" was there before too and this PR doesn't touch anything in regard to that, this will just enable us to install the level version of wine (3.5)? 🙏

ilovezfs commented:
>Why is it so important to get rid of the patch?

As a matter of policy, we require patches to at minimum be submitted upstream before inclusion (with some reasonable expectation that they're likely to be accepted), and to be removed if they're rejected, since Homebrew maintainers shouldn't be overriding upstream decisions to reject a patch nor should Homebrew maintainers be attempting to keep correctly rebasing a patch that will never be merged.

fxcoudert commented:
Maybe to phrase it somewhat differently from @ilovezfs: Homebrew's goal is to distribute software, not port it or maintain it.

We ship software that works on macOS, that has stable releases and is maintained. When something breaks, we'll encourage users reporting it upstream, so it gets fixed in the next version. In case of very serious breakage, we _may_ ship a patched version, _if_ it has been accepted by upstream (i.e. if they confirm their intent to fix it).

Wine's case is that they do not intend to fix the bug, and make their software work on macOS. Thus, one possible way out is to simply stop distributing it. It would be a shame, but we don't really want to become de facto maintainers of Wine on macOS.

**Possible ways forward:**
- investigate the bug more, to determine whether it truly is a macOS defect or a Wine issue
- create a fork of wine for macOS, include the patch, and we will consider distributing it under the rules for new formulas

fxcoudert commented:
The wine bug report at https://bugs.winehq.org/show_bug.cgi?id=34166#c99 hints that it may be an issue of `kCGLPFABackingStore` not being set. Here is a patch that unconditionally sets it for all OpenGL contexts in the macOS driver: https://gist.github.com/fxcoudert/539dc389d5f651bf713f684d43bb254b

Can someone try the patch (applying the patch **instead** of the current one) and report whether it fixes the issue? That would give some indication as to where the bug is, exactly.

ilovezfs commented:
>I changed the suggested patch to have:

@benjolitz How did you make this change? Can you confirm the file is still actually changed in your working tree? What is the current output of

```
cat $(brew --repo homebrew/core)/.git/config
git -C $(brew --repo homebrew/core) diff
```

benjolitz commented:
```
(cpython36) InvincibleReason:~$ cat $(brew --repo homebrew/core)/.git/config
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
	ignorecase = true
	precomposeunicode = true
	autocrlf = false
[remote "origin"]
	url = https://github.com/Homebrew/homebrew-core
	fetch = +refs/heads/master:refs/remotes/origin/master
[branch "master"]
	remote = origin
	merge = refs/heads/master
[homebrew]
	private = false
(cpython36) InvincibleReason:~$ git -C $(brew --repo homebrew/core) diff
diff --git a/Formula/wine.rb b/Formula/wine.rb
index b7477a45..8f128eef 100644
--- a/Formula/wine.rb
+++ b/Formula/wine.rb
@@ -32,6 +32,24 @@ class Wine < Formula
     sha256 "cceea27c906e9e5e7916281a57c36872c943802e998fede387fa7ccbc9d297ac" => :sierra
     sha256 "309e0b2978bb79ea627caba23d7f3bbb0cb54d3a6b08836b2cde387a743dae50" => :el_capitan
   end
+
+  devel do
+    url "https://dl.winehq.org/wine/source/3.x/wine-3.5.tar.xz"
+    mirror "https://downloads.sourceforge.net/project/wine/Source/wine-3.5.tar.xz"
+    sha256 "f42206143b7caf5527dbcd54c21f9b0a22338d8841d7fa08aa6c5f0b9dd72143"
+
+    # Patch to fix screen-flickering issues. Still relevant on 3.5.
+    # https://bugs.winehq.org/show_bug.cgi?id=34166
+    patch do
+      url "https://gist.githubusercontent.com/fxcoudert/539dc389d5f651bf713f684d43bb254b/raw/fd323ddc27df26b5ab7bebe433261a09525aeb0f/wine-mac.patch"
+      sha256 "e25e061b6bbb4b95cf3d1baae537b13c5a2508bcf329c48697cf83fe4f51d3b4"
+    end
+
+    resource "mono" do
+      url "https://dl.winehq.org/wine/wine-mono/4.7.1/wine-mono-4.7.1.msi", :using => :nounzip
+      sha256 "2c8d5db7f833c3413b2519991f5af1f433d59a927564ec6f38a3f1f8b2c629aa"
+    end
+  end

   head do
     url "https://source.winehq.org/git/wine.git"
(cpython36) InvincibleReason:~$
```

benjolitz commented:
If someone has a known binary that exhibits flickering, it would be nice to prove that the patch above works as intended. I'm happy to help out in testing.

fxcoudert commented:
@benjolitz I thought you tested it and reported that it did fixed the flickering on Black Mesa and Dishonored?

benjolitz commented:
I did not encounter any flickering. Like any software developer, I don’t believe my testing is comprehensive enough and chose to make a public request for more examples to verify against.

benjolitz commented:
To clarify, I tested the games I reported for a fair amount of time and didn’t see any flickering, tearing or other indication of the bug as reported in Wine’s bug tracker.

What I do not have is binaries of the original affected assets as cited in the Wine bug report, which would serve as an excellent test case.

fxcoudert commented:
@benjolitz can you test the same games, after re-installing wine without any patch, to confirm whether they display flickering?

benjolitz commented:
Yes. I've done additional testing since yesterday. I am sorry to say that my initial reports were in error.  TL;DR - I can't reproduce the problem at all, even with wine-stable.

## Hardware

Hardware: MacBook Pro (15-inch, 2017) @ 16GiB memory
Radeon Pro 560 4096 MB
Intel HD Graphics 630 1536 MB
`uname -r`: `17.5.0`

## Test Setup

I've tested wine-3.5 with and without the patch using:

- [Black Mesa (paid)](http://store.steampowered.com/app/362890/Black_Mesa/)
- [Dishonored (paid)](http://store.steampowered.com/app/205100/Dishonored/) 
- Steam
-  [3DMark03 (free)](https://www.futuremark.com/benchmarks/legacy)
- a [demo of Arx Fatalis (free)](http://store.steampowered.com/app/1700/Arx_Fatalis/)
- an [abandonware copy of Black and White (free)](https://www.myabandonware.com/game/black-white-a33). 
-  [CrystalMark 2009 (`CM09D2D.exe`)](https://crystalmark.info/download/archive/CrystalMark/CrystalMark2004R3.zip)

Patching was accomplished through editing the homebrew formulate directly (`brew edit wine`), saving and checking for presence or absence of the "Applying ... patch". 

After a test was complete, wine was removed using `brew remove wine --force` until no residual wine installations were left. To ensure that recent developments like CSMT are not interfering, I also ran tests with the DWORD for csmt set to `0`. 

To compare, I built wine-3.0 from source using `brew reinstall wine --build-from-source` without any patch. I use that on a fresh prefix formed using `rm -rf ~/.wine-3.0/ && (WINEARCH=win32 WINEPREFIX=~/.wine-3.0/ wineboot --init) && (WINEARCH=win32 WINEPREFIX=~/.wine-3.0/ winetricks fontsmooth=rgb) && (WINEARCH=win32 WINEPREFIX=~/.wine-3.0/ winecfg)`. 

Using Wine-3.0, (unpatched from source, patched with bottle) for:

-  [3DMark03 (free)](https://www.futuremark.com/benchmarks/legacy)
- an [abandonware copy of Black and White (free)](https://www.myabandonware.com/game/black-white-a33). 
-  [CrystalMark 2009 (`CM09D2D.exe`)](https://crystalmark.info/download/archive/CrystalMark/CrystalMark2004R3.zip)

## Results

The most obvious visual changes manifested in disabling CSMT (fonts would be corrupted in Black Mesa console, Black Mesa also displayed stuttering without CSMT enabled) - this issue occurred on both patched and unpatched clients if they were capable of CSMT.

I did not witness any graphical tearing, flickering or undesirable behavior in any of the reported programs above.

I expected to see some aberrations for 3DMark03 and/or Black&White but did not. One thing of note is that the bulk of affected clients appear to be running Macbook Airs - I wonder if there's something to that. I have also been unable to determine a post wine-2.x version that was tested/reported as affected by this bug. Only thing I can recommend is testing a Macbook Air on the free games/benchmarks I've cited above and/or reaching out to users who have publicly complained about this bug.

## References

[Fullscreen Flickering on OSX](https://bugs.winehq.org/show_bug.cgi?id=34166)

LinusU commented:
>  One thing of note is that the bulk of affected clients appear to be running Macbook Airs - I wonder if there's something to that.

I have a MacBook Air, I should be able to test, I have a copy black and white somewhere as well 🙌

LinusU commented:
Hmm, just tried out Black and White on my MacBook Air running High Sierra (10.13.4) using the official built wine 3.0 from winehq.com (which shouldn't have the no-flicker patch). It seems to be working without flickering.

I'm going to try it with Homebrew wine 3.0 without the patch next 👌

LinusU commented:
@AuroraWright when you made #23093 and updated from "Still relevant on 2.0." to "Still relevant on 3.0.", which version of macOS did you test with? and which program?

albinekb commented:
@LinusU maybe remove the patch for now and release the devel channel without the patch and we'll see if someone reports it? I can't repro the flicker either.

LinusU commented:
Rebased on master, removed the patch, and bumped the devel version to 3.6.

Hopefully the flickering is resolved by a recent macOS, and this is no longer a problem.

Cheers 🍻 

ping @ilovezfs

LinusU commented on Formula/wine.rb:
```diff
@@ -7,20 +7,13 @@
   desc "Run Windows applications without a copy of Microsoft Windows"
   homepage "https://www.winehq.org/"
-  revision 1
+  revision 2
```
This revision bump is because I'm removing the patch from stable (3.0)

ilovezfs commented:
I'd like to hear confirmation that the issue doesn't affect 10.11 or 10.12 before we remove the patch from stable.

LinusU commented:
That sounds smart 👍 

Do you want me to remove that commit from this PR, so that we can merge now, or do you want me to wait?

ilovezfs commented:
Yes, let's just merge devel sans patch and wait for some feedback :)

LinusU commented:
@ilovezfs fixed 👍

ilovezfs commented:
Thanks @LinusU!

LinusU commented:
Thank you, and everyone else in this thread, for all the help 💌
